core/insertion_marker_manager.ts

```diff
@@ -604,38 +599,16 @@ export class InsertionMarkerManager {
 
     const markerConn = this.markerConnection;
     const imBlock = markerConn.getSourceBlock();
-    const markerNext = imBlock.nextConnection;
     const markerPrev = imBlock.previousConnection;
     const markerOutput = imBlock.outputConnection;
 
-    const isNext = markerConn === markerNext;
-
-    const isFirstInStatementStack =
-        isNext && !(markerPrev && markerPrev.targetConnection);
-
-    const isFirstInOutputStack =
-        markerConn.type === ConnectionType.INPUT_VALUE &&
-        !(markerOutput && markerOutput.targetConnection);
-    // The insertion marker is the first block in a stack.  Unplug won't do
-    // anything in that case.  Instead, unplug the following block.
-    if (isFirstInStatementStack || isFirstInOutputStack) {
-      markerConn.targetBlock()!.unplug(false);
-    } else if (markerConn.type === ConnectionType.NEXT_STATEMENT && !isNext) {
-      // Inside of a C-block, first statement connection.
-      const innerConnection = markerConn.targetConnection;
-      if (innerConnection) {
-        innerConnection.getSourceBlock().unplug(false);
-      }
-
-      const previousBlockNextConnection =
-          markerPrev ? markerPrev.targetConnection : null;
-
-      imBlock.unplug(true);
-      if (previousBlockNextConnection && innerConnection) {
-        previousBlockNextConnection.connect(innerConnection);
-      }
+    if (!markerPrev?.targetConnection && !markerOutput?.targetConnection) {
```

Comment:
@rachel-fenichel I think this works, but could you PTAL and tell me if I'm missing anything?

Reply:
All of the complexity here is to handle a case that doesn't happen in Blockly but does happen in Scratch.

In Blockly if you drag a C-shaped block up a stack, it will attempt to insert, rather than wrap any of the blocks. When you get to the top of the stack, it'll offer to wrap the whole thing.
![38bd759f-301d-4e7e-871e-c31cf749daff](https://user-images.githubusercontent.com/13686399/224439310-ecc9690b-f6da-4df0-b5cb-32b5bc411188.gif)


In Scratch, it'll offer to wrap from the current position all the way to the end of the stack.

![ddd008ab-cd13-4b8b-ae21-97a16cec7982](https://user-images.githubusercontent.com/13686399/224439358-bc150c80-ee5e-4417-8a75-63e642faae9d.gif)

Disentangling the insertion marker when it's wrapping around blocks in the middle of the stack is handled by this code:
```
else if (markerConn.type === ConnectionType.NEXT_STATEMENT && !isNext) {
      // Inside of a C-block, first statement connection.
      const innerConnection = markerConn.targetConnection;
      if (innerConnection) {
        innerConnection.getSourceBlock().unplug(false);
      }

      const previousBlockNextConnection =
          markerPrev ? markerPrev.targetConnection : null;

      imBlock.unplug(true);
      if (previousBlockNextConnection && innerConnection) {
        previousBlockNextConnection.connect(innerConnection);
      }
    } 
```

This won't happen in base Blockly, because our implementation of `doDragChecks` in the `ConnectionChecker` doesn't allow it. But if you override `doDragChecks` you can definitely get back into this position, and Scratch's decision to do it was an active design choice, not an accident.

Reply:
Coolio, I'll write a unit test to document this behavior! and then see if I can implement this + still get the performance improvments.

Reply:
I found out that you can't achieve scratch-style insertion behavior behavior just by subclassing the connection checker and overriding `doDragChecks` :/ If you do you get this:

[Screen recording 2023-03-13 1.33.57 PM.webm](https://user-images.githubusercontent.com/25440652/224826433-d7b1e5d6-fc43-485f-84d3-bf5a5390b248.webm)

Scratch also overrides some logic in `connect` (which isn't possible in core blockly) to support inserting c-blocks. https://github.com/LLK/scratch-blocks/blob/66211340bc237a8fdc9017c8f2611768cb2fa3b4/core/connection.js#L150

I'm thinking that given this, it's best to keep the simplified insertion marker code, and file an issue to support scratch-style insertion. How does that sound to you @rachel-fenichel ?

Reply:
Sounds good. Link back to this conversation in the new issue, please.

Reply:
Filed https://github.com/google/blockly/issues/6898#issue-1626057110